### ISSUE TITLE
API call to require() module

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#696]
I propose an API call which would do the equivalent of `require("foo")` and push the exports object onto the stack.  In minisphere 4.0 I want to move to a more modular core API like Node.js has and it would be useful to be able to access core modules from C code.

**Comment by fatcerberus** [comment]
Thinking about this a bit more, an API which is just an interface for `require()` is a ticking time bomb security-wise: Even if you can guarantee you'll get the right module with a given name--which is not easy to guarantee, depending on how your module loader is set up--there's still the risk that some JS code will overwrite your `modSearch` and then all bets are off.

It's also not particularly useful in most cases to load C modules this way--if you already have the functionality implemented natively, you can just call the native functions directly.

What _would_ be useful, however, is the ability to compile/evaluate a given piece of code as though it were a CommonJS module, and push the `exports` to the stack.  This would allow use cases such as loading the TypeScript transpiler as a module directly from C (minisphere, e.g., loads this using `duk_eval()`, which pollutes the global object).

This could probably even be done without much footprint cost, since the "execute as module" primitive already needs to exist in order to have a built-in `require()`.

**Comment by svaarala** [comment]
Adding a C API for `require()` has been my intention also. When I wrote the original loader mechanism I intended to add it, but it's not as straightforward as just adding a `duk_require(ctx, id)` API call if one wants to support relative module resolution for C code too, so I had to postpone it. For example, if a native module is loaded, its initialization code may want to load more modules relative to its own module ID.

The API needs to handle that somehow, perhaps via an explicit resolution root argument or some hidden state. The explicit resolution root would be something like `duk_require(ctx, root, id)`. I'd prefer an explicit resolution root over hidden state (as is done for Ecmascript code out of necessity) because it's stateless and makes fewer assumptions about how the mechanism is used which is nice in the C API.

Re: overwriting Duktape.modSearch, that doesn't seem like a security issue to me:

- The code that could potentially overwrite Duktape.modSearch can also eval arbitrary code, edit Duktape.modLoaded directly to forcibly load/unload modules, etc. I don't think the effects of replacing Duktape.modSearch are any more serious.
- All code being loaded is ultimately controlled by the application. So if the application only provides trusted code via module loading, the only real issue would be accidental overwriting (which doesn't seem like an important issue to me). On the other hand, if the module loader loads untrusted code all the sandboxing issues come into play, and removing visibility of `Duktape.modLoaded` from application code is one of the sandboxing measures typically used.

Regardless, having "execute as a module" as a primitive would be useful and should handle ES6 modules too.

**Comment by fatcerberus** [comment]
For the security issues I was more referring to the case where some C code calls `duk_require()` expecting to get a known module back and ends up getting something else entirely because modSearch has been replaced.  It's unsafe for the same reason it's unsafe to do, say:

```c
duk_get_global_string(ctx, "Duktape");
duk_get_prop_string(ctx, -1, "act");
duk_call(ctx, 0);
```

Because modSearch can be replaced on the JS side (unless C code disables it by making the property nonwritable) the only safe time for native code to load a module by name is when it's doing so on behalf of Ecmascript code.  The native code can't safely make any assumptions about what it gets back.

If every line of JS code run by the app is trusted, this isn't an issue of course.  But that's not often the case - in my experience people use scripting engines because they want their app to be extensible. :)

**Comment by svaarala** [comment]
Yes, but what code will replace modSearch if all code is from a trusted source? And if not, there are a lot of issues to consider re: sandboxing if untrusted code (actively malicious rather than just broken) is going to be executed.

**Comment by svaarala** [comment]
Note the difference between accidents (say, a broken application extension) vs. actively malicious code. It seems highly unlikely to me that a module *accidentally* replaces modSearch. Actively malicious code can do that and much more besides, which is where the sandboxing issues (all of them) come into play. I don't see this API as a special case: it's as safe or unsafe as saying `require('foo')` from Ecmascript code, or `duk_eval_string(ctx, "require('foo')")`; why should a direct C API be any different?

**Comment by fatcerberus** [comment]
If all the code is trusted, there's no issue.  However that's often not the case - think platforms like Node.js or even minisphere.  In environments like that it would never be safe for C code to call `require`.

Maybe the code being run isn't even malicious, but has still replaced modSearch as part of implementing a custom module system.  in that case the C code might expect `duk_require()` to return one thing but it ends up returning something else (or throwing).  Here it's not a security issue, of course, it just prevents the user from replacing the module loader due to assumptions made by the native code.

**Comment by svaarala** [comment]
> If all the code is trusted, there's no issue. However that's often not the case - think platforms like Node.js or even minisphere. In environments like that it would never be safe for C code to call require.

But in such cases it's never safe to e.g. `eval`, `print`, `console.log`, `JSON.parse` etc: all of these can be replaced at any time by anything. Also stringifying values is unsafe because `Object.prototype.toString()` can be replaced, and so on.

I don't see why this API call is any different from these - or `duk_eval_string(ctx, "require('foo')")`.

**Comment by fatcerberus** [comment]
Agreed, which is why Duktape has e.g. `duk_json_decode()`. :)  I would never reference `JSON.parse()` directly from C for exactly that reason - it's not predictable.

I feel like I'm failing to articulate this properly - me typing on an iPhone likely has something to do with it.  I'll think on it some more.

**Comment by svaarala** [comment]
Hmm, so when you say:

> In environments like that it would never be safe for C code to call require.

... do you also mean that it's equally unsafe to use `require()` from Ecmascript in such environments?

In other words, you're actually more interested about loading a specific file bypassing the module cache rather than loading a module which would also be visible and shared by other callers? That would certainly be possible but the downside would be that the module wouldn't be cached or shared; if it were, it's exports table could again be modified by untrusted code.

By the way, if untrusted code is really an issue, it's pretty easy to hide the entire `Duktape` object from user code. This also makes the environment less Duktape specific. Or, if you need a few variables from the Duktape object, you can construct a replacement while the original retains a hidden reference and is used for the module cache.

**Comment by svaarala** [comment]
> I would never reference JSON.parse() directly from C for exactly that reason - it's not predictable.

Would the same apply to Ecmascript code which is part of the engine/platform? If not, why not? :)

But I think I see what you mean, and if I'm correct you're actually not really wanting to do a `require()` but rather load a module bypassing the module loader (including the search mechanism, caching, sharing, etc).

**Comment by svaarala** [comment]
Hmm, so suppose I load a module bypassing the module loader stuff and that module requires a sub-module. How should that get handled?

If that sub-require is loaded using the module loader, you'll have the same issue trusting the result (in your use case) because arbitrary stuff may be loaded.

If the sub-require is not handled by the module loader, how should it be handled?

Maybe in your use case the solution would be to have a separate "module loader domain" with a separate search function, separate module cache, which wouldn't be accessible to user code.

**Comment by fatcerberus** [comment]
Okay, let's stop using the word "untrusted" because that's a loaded term - despite what I said above I'm not actually concerned with security in this case.  My concern is more practical.  What it comes down to is that native code and Ecmascript may be written by different people.  Each one is a black box from the perspective of the other, mediated by a scripting language where monkey-patching is encouraged.  So to answer your question above:

> Would the same apply to Ecmascript code which is part of the engine/platform? If not, why not? :)

No, it would not.  If `JSON.parse` has been monkey-patched by Ecmascript code, then the Ecmascript code can deal with the fallout.  The native code, in the general case, cannot--and in fact has no way of even knowing the rug was pulled out from under it.

**Comment by fatcerberus** [comment]
Point taken on the submodule issue, though.  Unfortunately I don't really have any ideas how to handle that...

**Comment by fatcerberus** [comment]
Regarding hiding of the `Duktape` object: I briefly considered using `duk_del_prop_string()` to remove the object, but then I wondered if that would also prevent Duktape itself from accessing `Duktape.modSearch`.  Which actually highlights what I was getting at with the black box comment, now that I think of it...

**Comment by svaarala** [comment]
Maybe you're assuming Ecmascript code wouldn't be used for writing the infrastructure/platform parts (this is not generally true however)?

But in any case I think this is definitely not an issue for a `duk_require()` C API call, which I think should more or less just mimic `duk_eval_string(ctx, "require('foo')")` at least re: trusting modules. A separate mechanism for loading code in a module context may be needed in the future, as the function wrapper is a bit awkward and a little less efficient that necessary.

**Comment by svaarala** [comment]
> Regarding hiding of the Duktape object: I briefly considered using duk_del_prop_string() to remove the object, but then I wondered if that would also prevent Duktape itself from accessing Duktape.modSearch. Which actually highlights what I was getting at with the black box comment, now that I think of it...

That should be covered by the sandboxing documentation. Briefly, Duktape needs to maintain hidden references to various internal objects to satisfy Ecmascript semantics as many operations depend on the "original" value of certain built-ins. Internally these references are e.g. `thr->builtins[DUK_BIDX_ARRAY_PROTOTYPE]`.

The same applies to `Duktape.modSearch()`, the default require() implementation looks it up like this:

https://github.com/svaarala/duktape/blob/master/src/duk_bi_global.c#L1068-L1069

So yes, that will work even after you've deleted the visible `Duktape` object reference from the global object, as Duktape won't look up the object via the global object.

A bit off topic, this is one more example why I think the module loading mechanism should be more separated from Duktape core: for some environments it's very nice that Ecmascript code can implement a module search function so that it's easy to play with etc, while in others it'd make much more sense if the loading mechanism (including a search function, module cache, etc) would only be visible to C without having to go into actual sandboxing stuff.

**Comment by fatcerberus** [comment]
Right, as long as the native code and Ecmascript code are written by the same person (or in general, the same entity), there's no issue.  The minute you have a platform which can run arbitrary JS code though, all bets are off.  In that case it's not even safe for the infrastructure code written in _Ecmascript_ to access built-ins because the user code may have replaced them (with good intentions even!)

The main difference as far as I'm concerned is that it's easy for the startup code written in JS to simply do `var _JSON = JSON;` etc. before running any user code.  That's... well, quite a bit more tedious to do in C code.  Generally you'd have to use the stash for something like that.

**Comment by fatcerberus** [comment]
All that said, I do understand your point that it's not qualitatively different from the other "unsafe" cases and is just a consequence of using a scripting language where monkey-patching is encouraged.  So if you think there's a use case for `duk_require()`, I wouldn't be against it.

However in my case it's not what I need.  What I need is the ability to execute module code _as a module_ so that I can load, say, the TypeScript transpiler without making it visible through the global object in the process.  Currently I would need to create a separate heap for this, which is kind of overkill and somewhat inelegant. :)

**Comment by svaarala** [comment]
Right, but generally it's quite difficult to get any useful guarantees. For example, doing `var _JSON = JSON;` is not enough because user code can make even the stored JSON reference useless:

```
duk> Object.prototype.toJSON = function () { return 'foo'; };
= function anon() {/* ecmascript */}
duk> JSON.stringify([ 1, 2, 3])
= "foo"
```

So what I'm trying to say is that there are no actual guarantees from Duktape to prevent these (other than actual sandboxing) because there's no easily defined, useful, and consistent goal for such guarantees, at least that I can think of, without going into actual sandboxing.

**Comment by svaarala** [comment]
Agree, I think I understand what you'd like to accomplish, and in summary:

- Ability load code in a "module context" would be useful, though this doesn't solve the issue with sub-modules.
- Ability to have module loader domains which are kept separate would be useful.

As a first thought I don't think Duktape should try to provide a module loading mechanism with multiple security domains and such. It seems better to actually make fewer assumptions than currently but make it possible to write an application which has a module loader "on the side" if that requires Duktape support (e.g. the ability to load code in module context without writing the tedious wrappers yourself).

**Comment by fatcerberus** [comment]
Instead of multiple domains, what if `duk_eval_module()` (placeholder name) got as a parameter, a callback function which would receive the resolved module ID and return either the submodule source or exports on the valstack?  So basically, an impromptu modSearch.

**Comment by svaarala** [comment]
In addition to handling sub-requires there's also the issue of module caching/sharing which would need an additional module cache parameter. Also note that modules may not be require'd at load time so that callback would need to be in place indefinitely as would the cache. Adding these would essentially mean implementing the ability to have multiple module loaders in a single global environment, so I don't see easy (and version/feature stable) shortcuts.

**Comment by fatcerberus** [comment]
That's true.

As an initial implementation I would be okay with it if submodules were loaded using the normal modSearch (the transpiler has no external dependencies anyway),  we'd just have to be careful to do it in a way that doesn't prevent extensibility in the future. :)

**Comment by svaarala** [comment]
Right, and that's really my concern: adding a little bit of it makes it a versioning concern which needs a lot of cumulative work and distraction to support and back out of if it turns out to be wrong.

So I'd rather Duktape core not get involved in application module loading strategies (there are so many different approaches with quite different issues involved) except to provide the lowest level primitives needed to implement them. For example, you can implement a parallel module loader even currently as there's very little magic in the built-in loader (the function wrapper stuff can be used in the same way). I don't mind providing examples in the distributable of course.

**Comment by svaarala** [comment]
Maybe another issue to address in the 2.x module reword would be support for multiple module loaders in one global context. Though I'm not sure if there's much to address except to make sure there are no missing pieces in running multiple loaders in parallel.

There needs to be a module loader to handle `require()` and `import` from global code but this shouldn't prevent the application from using a separate loader e.g. for infrastructure components. So what's needed is the ability to initiate a module load using some module loader with that loader being inherited for sub-module loading. That doesn't require anything from Duktape for the CommonJS case because one can just create a new `require` for every module as Duktape does now (no direct support is needed from Duktape to do this). However, it may need some support for the ES6 module case because `import` is a keyword (unlike `require` which is a function binding).

**Comment by fatcerberus** [comment]
The main thing that has stopped me from rolling my own `require` so far is that the "magic" needed to support relative module IDs is tedious to implement in C (and it has to be in C because my engine is heavily sandboxed w.r.t. the file system).

**Comment by svaarala** [comment]
You could just copy the current implementation and modify it :)

**Comment by svaarala** [comment]
Anyway, I see your point. Maybe it could be implemented differently in the future, ES6 import will need something different in any case as it can't be replaced for each module being loaded.

Maybe Duktape could provide a "current module" context which both a CommonJS loader and ES6 loader could work with.

**Comment by fatcerberus** [comment]
I'll look later, but I'm assuming you attach the resolved ID directly to the `require` function somehow and then look that up when it's called.

**Comment by svaarala** [comment]
Yeah, it's stored in the `require` function and looked up using `duk_push_current_function()`.

Re: that module context, here's roughly what I mean. Suppose `foo/bar.js` is loaded and it has the module ID `foo/bar`:

```js
// foo/bar.js

var _currentModule = module;  // for later lookups

function test() {
    // Eventual sub-require when called.
    var sub = require('./quux');
}

exports.test = test;
```

When `require()` is called it could get access to the calling module using a `duk_push_calling_module()` or something thereof. That function would: (1) look up the calling function, (2) look up `_currentModule` within that function using a slow path lookup. The result would be the module table relevant to the calling function, which is the same for all functions declared in a module (regardless of nesting). The variable name would in an actual implementation be an internal one so that it'd be invisible as far as standard semantics are concerned.

A similar approach would be to attach a .module to each declared function, with the value being the same for all functions declared inside a module. But this would be property bloat compared to the alternative presented above.

**Comment by svaarala** [comment]
The above is (probably) technically incorrect if a module exports its require function which then gets called from someone outside the module:

```
// foo.js
exports.myRequire = require;
```

With the current require() implementation `require('foo').myRequire('./bar')` it will get resolved relative to `foo` as if the require() was inside the module. With the proposal above it would get resolved relative to the call site of `.myRequire()`.

This is probably not a practical concern though; and you can't export the "import context" of an ES6 module so it's a CommonJS issue only.

**Comment by fatcerberus** [comment]
Note that Node.js has `module.require()`:
https://nodejs.org/docs/latest/api/modules.html#modules_module_require_id

So there might be modules in the wild that rely on calling `require()` using another module's context, although I imagine the use cases would be limited and it doesn't apply to Duktape in its current state anyway.

**Comment by svaarala** [comment]
With the current anonymous function approach that case could be handled by prepending this to the module being loaded:

```js
module.require = function (id) { return require(id); };
```

That way the require call would always come from the module source itself.

This behavior is beyond CommonJS but still an interesting point of comparison. If the module context must follow `require` then it more or less has to be in the function itself, which also means a separate function needs to be created for every module loaded which is not ideal for low memory environments.

**Comment by fatcerberus** [comment]
I'd be curious to investigate how Node.js implements `require()`.  Given that it does have a separate `module.require` (instead of saying, just export `require` yourself), I wonder if it actually uses some kind of context tracking like you mentioned above for bare `require`...

**Comment by fatcerberus** [comment]
Interesting tidbit from the CommonJS spec:
> Relative identifiers are resolved relative to the identifier of the module **in which "require" is written and called.**

Emphasis mine.  Technically Duktape is out of spec right now.

**Comment by svaarala** [comment]
Interesting indeed :) I don't really understand that requirement completely. What would it mean, for example, to say:

```js
var foo = otherModule.exportedRequire;
foo();
```

`require` is not "written" here (assuming that part refers to the function name being called) but is called. That may seem like an odd distinction but `eval` semantics already do that to distinguish between direct and indirect eval calls.

I'm assuming that simply refers to the call site because anything else would impose requirements on the engine. Assuming it simply refers to the call site, the current Duktape implementation is then indeed violating the specification for an exported require function. The "_currentModule" approach I described above would actually fix that. Or maybe not completely; there are odd cases like calling through a `with` binding where the binding object is declared in yet another module, perhaps provided through a Proxy, etc, etc. Hardly real world cases though :)

**Comment by fatcerberus** [comment]
My assumption would be the same, that it refers to the call site.  Which explains the existence of `module.require` in Node - exporting `require` normally wouldn't work.

**Comment by svaarala** [comment]
I'll need to check how other CommonJS implementations actually behave, as even "call site" is not always that well defined, consider e.g. evals:

```js
var req = require('foo/bar').myRequire;

eval('req("./quux")');

var indirectEval = eval;
var globalObject = new Function('return this')();
globalObject.req = req;  // make available for indirect eval
indirectEval('req("./quux")');
```

A direct eval executes in the call site context so maybe that should resolve relative to the file where "eval" is called. On the other hand for Duktape eval code is a separate compilation unit. An indirect eval executes in the global context so maybe that shouldn't. These are not as trivial to implement (at least in Duktape) as just looking up the caller function.

Function constructor is also interesting:

```js
var req = require('foo/bar').myRequire;

var fun = new Function('r', 'r("./quux")');
fun(req);
```

Because the created function executes in the global scope it might be logical that resolution would also happen as if for global code.

Anyway, this is good to look into when things are changed around the module mechanism :)

**Comment by fatcerberus** [comment]
Hm, I guess this is no longer relevant now that module support has been moved out of Duktape core.  Move to close?

**Comment by svaarala** [comment]
Maybe not - there will be an ES6 import mechanism later on. With the module mechanism out of the core I think the cleanest way to implement that would be to provide a way to register the provider for require/ES6 import. And there should also be a C API to require/import a module. Not sure how it would work yet, but this is still relevant.
